### PR TITLE
core: Implement channel tracing for some events

### PR DESCRIPTION
**Pull request by dapengzhang0**
Includes 
- API plumbing for channel tracing `+ManagedChannelBuilder.maxTraceEvents()`
- trace logging for channel/subchannel creation events and name resolving events.

Trace logging for all other events will be added in subsequent PRs.

### CODE REVIEW

**Comment by dapengzhang0** [comment]
Rebasing to resolve conflicting files.

**Comment by dapengzhang0** [comment]
@zpencer PTAL

File: core/src/main/java/io/grpc/internal/Channelz.java [hunk 1]
```diff
@@ -525,17 +530,46 @@ private Event(
         checkArgument(
             channelRef == null || subchannelRef == null,
             "at least one of channelRef and subchannelRef must be null");
-        this.description = checkNotNull(description, "description");
+        this.description = description;
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/Channelz.java]: Validation is moved to builder, so not to check again here. I should have also moved the check for `channelRef == null || subchannelRef == null` in the builder as well.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 2]
```diff
@@ -470,11 +474,20 @@ public LogId getLogId() {
   public ListenableFuture<ChannelStats> getStats() {
     SettableFuture<ChannelStats> ret = SettableFuture.create();
     ChannelStats.Builder builder = new ChannelStats.Builder();
+
+    EquivalentAddressGroup addressGroupSnapshot;
+    List<WithLogId> transportsSnappshot;
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Done.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java [hunk 3]
```diff
@@ -2069,6 +2067,93 @@ public void channelsAndSubchannels_instrumented_name() throws Exception {
...
+    assertThat(getStats(channel).channelTrace.events).contains(new ChannelTrace.Event.Builder()
+        .setDescription("Child channel created")
+        .setSeverity(ChannelTrace.Event.Severity.CT_INFO)
+        .setTimestampNaonos(timer.getTicker().read())
+        .setSubchannelRef(subchannel.getInternalSubchannel())
+        .build());
+    assertThat(getStats(subchannel).channelTrace.events).contains(new ChannelTrace.Event.Builder()
+        .setDescription("Subchannel created")
+        .setSeverity(ChannelTrace.Event.Severity.CT_INFO)
+        .setTimestampNaonos(timer.getTicker().read())
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java]: Done.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 4]
```diff
@@ -574,11 +582,29 @@ public void execute(Runnable command) {
...
+    };
     channelCallTracer = callTracerFactory.create();
     this.channelz = checkNotNull(builder.channelz);
     channelz.addRootChannel(this);
+
+    maxTraceEvents = builder.maxTraceEvents;
+    if (maxTraceEvents > 0) {
+      long currentTimeNanos = timeProvider.currentTimeNanos();
+      channelTracer = new ChannelTracer(builder.maxTraceEvents, currentTimeNanos);
+      channelTracer.reportEvent(new ChannelTrace.Event.Builder()
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Done.

**Comment by dapengzhang0** [comment]
PTALA

File: core/src/main/java/io/grpc/internal/ChannelTracer.java [hunk 5]
```diff
@@ -39,7 +39,12 @@
   @GuardedBy("lock")
   private int eventsLogged;
 
-  ChannelTracer(final int maxEvents, long channelCreationTimeNanos) {
+  /**
+   * Creates a channel tracer and log the creation event of the underlying channel.
+   *
+   * @param channelType Chennel, Subchannel, or OobChannel
+   */
+  ChannelTracer(final int maxEvents, long channelCreationTimeNanos, String channelType) {
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ChannelTracer.java]: A nullcheck on channelType won't hurt

**Comment by dapengzhang0** [comment]
Thanks for the review!

**Pull request merged**